components/brave_wallet/browser/eth_data_parser.cc

```diff
@@ -80,66 +38,84 @@ bool GetTransactionInfoFromData(const std::string& data,
-    std::string address, value;
-    std::string left_over_data = data.substr(10);
-    // Intentional copy of left_over_data
-    if (!GetAddressArgFromData(std::string(left_over_data), &address,
-                               &left_over_data))
+  std::string selector = base::ToLowerASCII(data.substr(0, 10));
+  std::string left_over_data = data.substr(10);
+  if (selector == "0xa9059cbb") {
+    *tx_type = mojom::TransactionType::ERC20Transfer;
+    if (!brave_wallet::ABIDecode({"address", "uint256"}, left_over_data,
+                                 tx_params, tx_args))
       return false;
-    if (!GetUint256HexFromData(std::string(left_over_data), &value,
-                               &left_over_data))
-      return false;
-    // Very strictly must have correct data
-    if (left_over_data.length() > 0)
-      return false;
-
-    if (tx_args) {
-      tx_args->push_back(address);
-      tx_args->push_back(value);
-    }
-
-    if (tx_params) {
-      tx_params->push_back("address");
-      tx_params->push_back("uint256");
-    }
-  } else if (*tx_type == mojom::TransactionType::ERC721TransferFrom ||
-             *tx_type == mojom::TransactionType::ERC721SafeTransferFrom) {
-    std::string from, to, token_id;
-    std::string left_over_data = data.substr(10);
-    // Intentional copy of left_over_data
-    if (!GetAddressArgFromData(std::string(left_over_data), &from,
-                               &left_over_data))
+  } else if (selector == "0x095ea7b3") {
+    *tx_type = mojom::TransactionType::ERC20Approve;
+    if (!brave_wallet::ABIDecode({"address", "uint256"}, left_over_data,
+                                 tx_params, tx_args))
       return false;
-    if (!GetAddressArgFromData(std::string(left_over_data), &to,
-                               &left_over_data))
+  } else if (selector == "0x23b872dd") {
+    *tx_type = mojom::TransactionType::ERC721TransferFrom;
+    if (!brave_wallet::ABIDecode({"address", "address", "uint256"},
+                                 left_over_data, tx_params, tx_args))
       return false;
-    if (!GetUint256HexFromData(std::string(left_over_data), &token_id,
-                               &left_over_data))
+  } else if (selector == "0x42842e0e") {
+    *tx_type = mojom::TransactionType::ERC721SafeTransferFrom;
+    if (!brave_wallet::ABIDecode({"address", "address", "uint256"},
+                                 left_over_data, tx_params, tx_args))
       return false;
-    // Very strictly must have correct data
-    if (left_over_data.length() > 0)
+  } else if (selector == "0x3598d8ab") {
+    // Function:
+    // sellEthForTokenToUniswapV3(bytes encodedPath,
+    //                            uint256 minBuyAmount,
+    //                            address recipient)
+    *tx_type = mojom::TransactionType::ETHSwap;
+    if (!brave_wallet::ABIDecode({"bytes", "uint256", "address"},
+                                 left_over_data, tx_params, tx_args))
```

Comment:
Where does this made-up API data get generated?

Reply:
Here's the ABI I'm parsing: https://www.4byte.directory/signatures/?bytes4_signature=0x3598d8ab. It is one of the common 0x contract calls for swap, among a few others.

I was able to figure out the format of `bytes encodedPath` by going through [0x contract tests](https://github.com/0xProject/protocol/blob/901d400d629e277fe24b0eac2b290b69a05a3a0d/contracts/zero-ex/test/features/uniswapv3_test.ts#L195-L197).

I'm then organising the parsed info in the following format, which is common for all `ETHSwap` transactions.

```
{
    "bytes",    // fill path (maker asset, followed by hops if any, followed by taker asset)
    "uint256",  // maker amount
    "uint256"   // taker amount
}
```

The idea is to represent all the various 0x contract calls using a common `ETHSwap` type, and we can then build a good UX with all the parsed information.

Reply:
I see, maybe add comment with [interface link ](https://github.com/0xProject/protocol/blob/b46eeadc64485288add5940a210e1a7d0bcb5481/contracts/zero-ex/contracts/src/features/interfaces/IUniswapV3Feature.sol#L34) around these lines saying it's from 0x because it is straightforward to me that [`transformERC20`](https://protocol.0x.org/en/latest/advanced/erc20_transformations.html#erc20-transformations) is from 0x but not this one

Reply:
Improved the code comments, and extracted out the logic to parse the Uniswap-encoded bytes into a separate function in `eth_abi_decoder.cc`.

See 198f58536305de7168311aa716138314b0ca1f3b.